lib/inspec/config.rb

```diff
@@ -0,0 +1,286 @@
+        unless valid_fields.include?(seen_field)
+          raise Inspec::ConfigError::Invalid, "Unrecognized top-level configuration field #{seen_field}.  Recognized fields: #{valid_fields.join(', ')}"
+        end
+      end
+    end
+
+    def validate_reporters(reporters)
+      return if reporters.nil?
+      # TODO: move this into a reporter plugin type system
+      valid_types = [
+        'automate',
+        'cli',
+        'documentation',
+        'html',
+        'json',
+        'json-automate',
+        'json-min',
+        'json-rspec',
+        'junit',
+        'progress',
+        'yaml',
+      ]
+
+      reporters.each do |reporter_name, reporter_config|
+        raise NotImplementedError, "'#{reporter_name}' is not a valid reporter type." unless valid_types.include?(reporter_name)
+
+        next unless reporter_name == 'automate'
+        %w{token url}.each do |option|
+          raise Inspec::ReporterError, "You must specify a automate #{option} via the config file." if reporter_config[option].nil?
+        end
+      end
+
+      # check to make sure we are only reporting one type to stdout
+      stdout_reporters = 0
+      reporters.each_value do |reporter_config|
+        stdout_reporters += 1 if reporter_config['stdout'] == true
+      end
+
+      raise ArgumentError, 'The option --reporter can only have a single report outputting to stdout.' if stdout_reporters > 1
+    end
+
+    #-----------------------------------------------------------------------#
+    #                         Merging Options
+    #-----------------------------------------------------------------------#
+    def merge_options
+      options = Thor::CoreExt::HashWithIndifferentAccess.new({})
+
+      # Lowest precedence: default, which may vary by command
+      options.merge!(@defaults)
+
+      # Middle precedence: merge in any CLI options defined from the config file
+      options.merge!(config_file_cli_options)
+      # Reporter options may be defined top-level.
+      options.merge!(config_file_reporter_options)
+
+      # Highest precedence: merge in any options defined via the CLI
+      options.merge!(@cli_opts)
+
+      options
+    end
+
+    #-----------------------------------------------------------------------#
+    #                          Finalization
+    #-----------------------------------------------------------------------#
+    def finalize_options
+      options = @merged_options.dup
+
+      finalize_set_top_level_command(options)
+      finalize_parse_reporters(options)
+      finalize_handle_sudo(options)
+      finalize_compliance_login(options)
+
+      Thor::CoreExt::HashWithIndifferentAccess.new(options)
+    end
+
+    def finalize_set_top_level_command(options)
+      options[:type] = @command_name
+      Inspec::BaseCLI.inspec_cli_command = @command_name # TODO: move to a more relevant location
+    end
+
+    def finalize_parse_reporters(options) # rubocop:disable Metrics/AbcSize
```

Comment:
There's a lot going on here, I think this is covered end-to-end in your unit tests, but I'd like to see this method isolated and tests created for it for the sake of regression prevention.

I also feel you could split this into a few methods based on Hash or Array which might make it a bit easier to digest (and avoid having to disable the AbcSize cop)

Reply:
This is pretty much directly copied out of base_cli.rb. I don't claim to make any changes to the code, nor improve the testing on it.

https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/lib/inspec/base_cli.rb#L136

Reply:
Ah, I was thinking about the Boy Scout Rule you've mentioned in the past. Does it only apply to small things (I recognize adding tests for this would be significant work...at least for me)

Reply:
It's a rule of thumb, I guess, and relies on intuition.

The goal of this PR series is to get the config reading and validation into a proper class, and then expose a simpler way of reading credentials. So, I must copy a lot of code from where it was (scattered as class methods in Inspec::BaseCLI) to its new home. (On the 2nd PR of the series, I delete it from its previous location - which perhaps I should have done on this PR). In order to help people follow the changes, I'm just doing simple copying as much as possible. Making improvements (like adding tests, renaming vars, etc) would add noise to the PR.

We could improve testing of the reporter parser on a later PR, either on this series, or treat that as a separate issue - I think that might take some real effort to really exercise that facility.  And the problem with adding tests, you know... you may find bugs, which then you must fix! Which makes me thing adding tests for the reporter parser is probably best as a separate issue.